src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -486,9 +498,21 @@ protected override bool MoveNextCore()
                         ds = ParquetReader.Read(_loader._parquetStream, _loader._parquetOptions, _readerOptions);
                     }
 
-                    int[] dataSetOrder = _rand == null ? Utils.GetIdentityPermutation(ds.RowCount) : Utils.GetRandomPermutation(_rand, ds.RowCount);
-                    _dataSetEnumerator = dataSetOrder.GetEnumerator();
-                    _curDataSetRow = dataSetOrder[0];
+                    try
+                    {
+                        int[] dataSetOrder = _rand == null ? Utils.GetIdentityPermutation(ds.RowCount) : Utils.GetRandomPermutation(_rand, ds.RowCount);
+                        _dataSetEnumerator = dataSetOrder.GetEnumerator();
+                        _curDataSetRow = dataSetOrder[0];
+                    }
+                    catch (Exception e)
+                    {
+                        if (e is OutOfMemoryException)
+                        {
+                            throw new InvalidDataException("Error caused because block size too big. Try decreasing block size.", e);
```

Comment:
>InvalidDataException [](start = 38, length = 20)

Just an FYI: `InvalidDataException`'s '[documentation](https://msdn.microsoft.com/en-us/library/system.io.invaliddataexception(v=vs.110).aspx) says "The exception that is thrown when a data stream is in an invalid format." This does not describe the scenario here: your data is not in an invalid format, rather, something is wrong with our implementation to interpret that format.

Reply:
Also as @glebuk says, might be best to use `_host.Except` here... or `_host.ExceptParam` if you can isolate it to a specific user-provided parameter that was somehow "wrong."